pkg/controller/attach_detach.go

```diff
@@ -312,64 +311,69 @@ func (r *ReconcileAttachDetach) triggerDetach(ctx context.Context, azVolumeAttac
-			defer cloudCancel()
-
-			var updateFunc func(obj client.Object) error
-			updateMode := azureutils.UpdateCRIStatus
-			detachErr = r.detachVolume(cloudCtx, azVolumeAttachment.Spec.VolumeID, azVolumeAttachment.Spec.NodeName)
-			if detachErr != nil {
-				updateFunc = func(obj client.Object) error {
-					azv := obj.(*azdiskv1beta2.AzVolumeAttachment)
-					azv = updateError(azv, detachErr)
-					_, derr := updateState(azv, azdiskv1beta2.DetachmentFailed, forceUpdate)
-					return derr
-				}
-				//nolint:contextcheck // final status update of the CRI must occur even when the current context's deadline passes.
-				if _, uerr := azureutils.UpdateCRIWithRetry(goCtx, nil, r.cachedClient, r.azClient, azVolumeAttachment, updateFunc, consts.ForcedUpdateMaxNetRetry, updateMode); uerr != nil {
-					w.Logger().Errorf(uerr, "failed to update final status of AzVolumeAttachement")
-				}
-			} else {
-				//nolint:contextcheck // delete of the CRI must occur even when the current context's deadline passes.
-				if derr := r.cachedClient.Delete(goCtx, azVolumeAttachment); derr != nil {
-					w.Logger().Error(derr, "failed to delete AzVolumeAttachment")
-				}
-			}
-		}()
-		<-waitCh
+	var updatedObj client.Object
+	if updatedObj, err = azureutils.UpdateCRIWithRetry(ctx, nil, r.cachedClient, r.azClient, azVolumeAttachment, updateFunc, consts.NormalUpdateMaxNetRetry, azureutils.UpdateCRIStatus); err != nil {
+		return err
 	}
+	azVolumeAttachment = updatedObj.(*azdiskv1beta2.AzVolumeAttachment)
+
+	w.Logger().V(5).Info("Detaching volume")
+	waitCh := make(chan goSignal)
+	//nolint:contextcheck // call is asynchronous; context is not inherited by design
+	go func() {
+		var detachErr error
+		_, goWorkflow := workflow.New(ctx)
+		defer func() { goWorkflow.Finish(detachErr) }()
+		waitCh <- goSignal{}
+
+		goCtx := goWorkflow.SaveToContext(context.Background())
+		cloudCtx, cloudCancel := context.WithTimeout(goCtx, cloudTimeout)
+		defer cloudCancel()
+
+		var updateFunc func(obj client.Object) error
+		updateMode := azureutils.UpdateCRIStatus
+		detachErr = r.detachVolume(cloudCtx, azVolumeAttachment.Spec.VolumeID, azVolumeAttachment.Spec.NodeName)
+		if detachErr != nil {
+			updateFunc = func(obj client.Object) error {
+				azv := obj.(*azdiskv1beta2.AzVolumeAttachment)
+				_, derr := reportError(azv, azdiskv1beta2.DetachmentFailed, detachErr)
+				return derr
+			}
+			//nolint:contextcheck // final status update of the CRI must occur even when the current context's deadline passes.
+			if _, uerr := azureutils.UpdateCRIWithRetry(goCtx, nil, r.cachedClient, r.azClient, azVolumeAttachment, updateFunc, consts.ForcedUpdateMaxNetRetry, updateMode); uerr != nil {
+				w.Logger().Errorf(uerr, "failed to update final status of AzVolumeAttachement")
+			}
+		} else {
+			updateFunc = func(obj client.Object) error {
+				azv := obj.(*azdiskv1beta2.AzVolumeAttachment)
+				_, derr := updateState(azv, azdiskv1beta2.Detached, normalUpdate)
```

Comment:
Why are we updating the state to `Detached` here since we're immediately marking it for delete? (I have a vague recollection of some discussion, but I forget the details.)

Reply:
This update is to prevent orphaned `AzVolumeAttachment` CRIs mentioned in the [comment](https://github.com/kubernetes-sigs/azuredisk-csi-driver/pull/1570#discussion_r1017012162).

Reply:
Let's seek different way to avoid orphaned `AzVolumeAttachment` scenario as this will add to our already heavy apiserver load.

Reply:
How about removing finalizer before making the [DELETE request](https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/edd3cb8795b269297cc68820418830b4eab43623/pkg/controller/attach_detach.go#L371)?

Reply:
Both the `replica` and `attach_detach` controllers can delete an `AzVolumeAttachment`. The finalizer is removed by the `attach_detach` controller's `Reconcile` function and allows the `replica` controller to coordinate on the deletion of a failed replica attachment. At least, this is my understanding of the flow.

@sunpa93 Would it work for both the `replica` and `attach_detach` controllers to remove the finalizer and delete the `AzVolumeAttachment` objects as needed?

Reply:
Confirmed with Sung Jun, it seems like changing definition of `isOperationInProcess` function back to `return target.Status.State == azdiskv1beta2.Attaching || (target.Status.State == azdiskv1beta2.Detaching && !deleteRequested)` can solve this issue.